lib/render.js

```diff
@@ -44,40 +45,81 @@ function applyRenderHooks(ref, data, locals) {
   }
 
   localsClone = _.cloneDeep(locals);
+
   return bluebird.reduce(setup.plugins, (data, plugin) => {
     return plugin.render(ref, data, localsClone);
   }, data);
 }
 
 /**
- * Take in the data returned from Amphora, call `render` hooks,
- * and return HTML output.
+ * Given the state, find the root component, compose
+ * the data into a structure Kiln will need and then
+ * render the HTML
  *
- * @param  {Object} data
- * @return {Promise<String>} html
+ * @param  {Object} state
+ * @return {Function<String>}
  */
-function render(data) {
-  return applyRenderHooks(data._layoutRef, data._data, data.locals)
-  .then(_data => {
-    data._data = _data;
+function makeHtml(state) {
+  return function (_data) {
+    state._data = _data;
 
-    const template = clayUtils.getComponentName(data._layoutRef || data._self),
+    const template = clayUtils.getComponentName(state._layoutRef || state._self),
       rootPartial = hbs.partials[template],
-      composedData = composeData(data);
+      composedData = composeData(state);
 
     if (!rootPartial) {
-      throw new Error(`Missing template for ${data.template}`);
+      throw new Error(`Missing template for ${template}`);
     }
 
     return rootPartial(composedData);
-  })
-  .then(mediaService.append(data._media, data.locals.edit, data.locals.site))
-  .then(result => {
-    return {
-      output: result,
+  };
+}
+
+/**
+ * 1. Apply render hooks
+ * 2. Get the HTML string
+ * 3. Add in scripts/styles
+ * 4. Terminate the response
+ * 5. Log the status
+ *
+ * @param  {Object} state
+ * @param  {Object} res
+ * @return {Promise}
+ */
+function render(state, res) {
+  const hrStart = process.hrtime();
+
+  return applyRenderHooks(state._layoutRef || state.self, state._data, state.locals)
+    .then(makeHtml(state))
+    .then(mediaService.injectScriptsAndStyles(state))
+    .then(result => {
+      res.type('text/html');
+      res.send(result);
+      return true; // Return true for logging statement
```

Comment:
The fnc `logTime` returns doesn't have any arguments so we don't need to return anything here

Reply:
Oh yeah, forgot about that one